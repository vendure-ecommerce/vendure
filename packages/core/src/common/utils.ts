import { AssetType } from '@vendure/common/lib/generated-types';
import { ID } from '@vendure/common/lib/shared-types';
import { Observable, Observer } from 'rxjs';

/**
 * Takes a predicate function and returns a negated version.
 */
export function not(predicate: (...args: any[]) => boolean) {
    return (...args: any[]) => !predicate(...args);
}

/**
 * Returns a predicate function which returns true if the item is found in the set,
 * as determined by a === equality check on the given compareBy property.
 */
export function foundIn<T>(set: T[], compareBy: keyof T) {
    return (item: T) => set.some(t => t[compareBy] === item[compareBy]);
}

/**
 * Identity function which asserts to the type system that a promise which can resolve to T or undefined
 * does in fact resolve to T.
 * Used when performing a "find" operation on an entity which we are sure exists, as in the case that we
 * just successfully created or updated it.
 */
export function assertFound<T>(promise: Promise<T | undefined>): Promise<T> {
    return promise as Promise<T>;
}

/**
 * Compare ID values for equality, taking into account the fact that they may not be of matching types
 * (string or number).
 */
export function idsAreEqual(id1?: ID, id2?: ID): boolean {
    if (id1 === undefined || id2 === undefined) {
        return false;
    }
    return id1.toString() === id2.toString();
}

/**
 * Returns the AssetType based on the mime type.
 */
export function getAssetType(mimeType: string): AssetType {
    const type = mimeType.split('/')[0];
    switch (type) {
        case 'image':
            return AssetType.IMAGE;
        case 'video':
            return AssetType.VIDEO;
        default:
            return AssetType.BINARY;
    }
}

/**
 * A simple normalization for email addresses. Lowercases the whole address,
 * even though technically the local part (before the '@') is case-sensitive
 * per the spec. In practice, however, it seems safe to treat emails as
 * case-insensitive to allow for users who might vary the usage of
 * upper/lower case. See more discussion here: https://ux.stackexchange.com/a/16849
 */
export function normalizeEmailAddress(input: string): string {
    return input.trim().toLowerCase();
}

/**
 * Converts a value that may be wrapped into a Promise or Observable into a Promise-wrapped
 * value.
 */
export async function awaitPromiseOrObservable<T>(value: T | Promise<T> | Observable<T>): Promise<T> {
    let result = await value;
    if (result instanceof Observable) {
        result = await result.toPromise();
    }
    return result;
}

/**
 * @description
 * Returns an observable which executes the given async work function and completes with
 * the returned value. This is useful in methods which need to return
 * an Observable but also want to work with async (Promise-returning) code.
 *
 * @example
 * ```TypeScript
 * \@Controller()
 * export class MyWorkerController {
 *
 *     \@MessagePattern('test')
 *     handleTest() {
 *         return asyncObservable(async observer => {
 *             const value = await this.connection.fetchSomething();
 *             return value;
 *         });
 *     }
 * }
 * ```
 */
export function asyncObservable<T>(work: (observer: Observer<T>) => Promise<T | void>): Observable<T> {
    return new Observable<T>(subscriber => {
        (async () => {
            try {
                const result = await work(subscriber);
                if (result) {
                    subscriber.next(result);
                }
                subscriber.complete();
            } catch (e) {
                subscriber.error(e);
            }
        })();
    });
}
