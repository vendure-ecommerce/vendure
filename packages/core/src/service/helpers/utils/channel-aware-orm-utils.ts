import { ID, Type } from '@vendure/common/lib/shared-types';
import { Connection, FindManyOptions, FindOptionsUtils } from 'typeorm';

import { ChannelAware } from '../../../common/types/common-types';
import { VendureEntity } from '../../../entity';

/**
 * Like the TypeOrm `Repository.findByIds()` method, but limits the results to
 * the given Channel.
 */
export function findByIdsInChannel<T extends ChannelAware | VendureEntity>(
    connection: Connection,
    entity: Type<T>,
    ids: ID[],
    channelId: ID,
    findOptions?: FindManyOptions<T>,
    eager = true,
) {
    //the syntax described in https://github.com/typeorm/typeorm/issues/1239#issuecomment-366955628
    //breaks if the array is empty
    if(ids.length === 0){
        return Promise.resolve([]);
    }
    
    const qb = connection.getRepository(entity).createQueryBuilder('product');
    FindOptionsUtils.applyFindManyOptionsOrConditionsToQueryBuilder(qb, findOptions);
    if (eager) {
        // tslint:disable-next-line:no-non-null-assertion
        FindOptionsUtils.joinEagerRelations(qb, qb.alias, qb.expressionMap.mainAlias!.metadata);
    }
    return qb
        .leftJoin('product.channels', 'channel')
        .andWhere("product.id IN (:...ids)", { ids })
        .andWhere('channel.id = :channelId', { channelId })
        .getMany();
}

/**
 * Like the TypeOrm `Repository.findOne()` method, but limits the results to
 * the given Channel.
 */
export function findOneInChannel<T extends ChannelAware | VendureEntity>(
    connection: Connection,
    entity: Type<T>,
    id: ID,
    channelId: ID,
    findOptions?: FindManyOptions<T>,
    eager = true,
) {
    const qb = connection.getRepository(entity).createQueryBuilder('product');
    FindOptionsUtils.applyFindManyOptionsOrConditionsToQueryBuilder(qb, findOptions);
    if (eager) {
        // tslint:disable-next-line:no-non-null-assertion
        FindOptionsUtils.joinEagerRelations(qb, qb.alias, qb.expressionMap.mainAlias!.metadata);
    }
    return qb
        .leftJoin('product.channels', 'channel')
        .andWhere('product.id = :id', { id })
        .andWhere('channel.id = :channelId', { channelId })
        .getOne();
}
