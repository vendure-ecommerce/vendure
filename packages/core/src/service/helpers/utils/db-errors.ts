/**
 * Returns true if the given error represents a foreign key constraint violation
 * across supported drivers (Postgres, MySQL/MariaDB, SQLite).
 */
export function isForeignKeyViolationError(e: unknown): boolean {
    const err: any = e || {};
    const code = err.code ?? err.driverError?.code ?? err.errno ?? err.driverError?.errno;

    // Postgres: 23503, MySQL/MariaDB: 1451/1452, SQLite: SQLITE_CONSTRAINT_FOREIGNKEY,
    if (code === '23503' || code === 1451 || code === 1452 || code === 'SQLITE_CONSTRAINT_FOREIGNKEY') {
        return true;
    }
    const msg = String(err.message ?? err.driverError?.message ?? '');
    return /\bforeign key\b/i.test(msg);
}
