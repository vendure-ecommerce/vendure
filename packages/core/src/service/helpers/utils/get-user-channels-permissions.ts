import { Permission } from '@vendure/common/lib/generated-types';
import { ID } from '@vendure/common/lib/shared-types';
import { unique } from '@vendure/common/lib/unique';

import { Role } from '../../../entity/role/role.entity';
import { User } from '../../../entity/user/user.entity';

export interface UserChannelPermissions {
    id: ID;
    token: string;
    code: string;
    permissions: Permission[];
}

/**
 * Returns an array of Channels and permissions on those Channels for the given User.
 */
export function getUserChannelsPermissions(user: User): UserChannelPermissions[] {
    return getChannelPermissions(user.roles);
}

/**
 * @description
 * Returns an array of Channels and permissions on those Channels for the given Roles.
 */
export function getChannelPermissions(roles: Role[]): UserChannelPermissions[] {
    const channelsMap: { [code: string]: UserChannelPermissions } = {};

    for (const role of roles) {
        for (const channel of role.channels) {
            if (!channelsMap[channel.code]) {
                channelsMap[channel.code] = {
                    id: channel.id,
                    token: channel.token,
                    code: channel.code,
                    permissions: [],
                };
            }
            channelsMap[channel.code].permissions = unique([
                ...channelsMap[channel.code].permissions,
                ...role.permissions,
            ]);
        }
    }

    return Object.values(channelsMap).sort((a, b) => (a.id < b.id ? -1 : 1));
}
