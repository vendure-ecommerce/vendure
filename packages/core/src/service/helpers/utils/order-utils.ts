import { summate } from '@vendure/common/lib/shared-utils';

import { OrderItem } from '../../../entity/order-item/order-item.entity';
import { Order } from '../../../entity/order/order.entity';
import { PaymentState } from '../payment-state-machine/payment-state';

/**
 * Returns true if the Order total is covered by Payments in the specified state.
 */
export function orderTotalIsCovered(order: Order, state: PaymentState | PaymentState[]): boolean {
    const paymentsTotal = totalCoveredByPayments(order, state);
    return paymentsTotal === order.totalWithTax;
}

/**
 * Returns the total amount covered by all Payments (minus any refunds)
 */
export function totalCoveredByPayments(order: Order, state?: PaymentState | PaymentState[]): number {
    const payments = state
        ? Array.isArray(state)
            ? order.payments.filter(p => state.includes(p.state))
            : order.payments.filter(p => p.state === state)
        : order.payments.filter(
              p => p.state !== 'Error' && p.state !== 'Declined' && p.state !== 'Cancelled',
          );
    let total = 0;
    for (const payment of payments) {
        const refundTotal = summate(payment.refunds, 'total');
        total += payment.amount - Math.abs(refundTotal);
    }
    return total;
}

/**
 * Returns true if all (non-cancelled) OrderItems are delivered.
 */
export function orderItemsAreDelivered(order: Order) {
    return getOrderItems(order)
        .filter(orderItem => !orderItem.cancelled)
        .every(isDelivered);
}

/**
 * Returns true if at least one, but not all (non-cancelled) OrderItems are delivered.
 */
export function orderItemsArePartiallyDelivered(order: Order) {
    const nonCancelledItems = getNonCancelledItems(order);
    return nonCancelledItems.some(isDelivered) && !nonCancelledItems.every(isDelivered);
}

/**
 * Returns true if at least one, but not all (non-cancelled) OrderItems are shipped.
 */
export function orderItemsArePartiallyShipped(order: Order) {
    const nonCancelledItems = getNonCancelledItems(order);
    return nonCancelledItems.some(isShipped) && !nonCancelledItems.every(isShipped);
}

/**
 * Returns true if all (non-cancelled) OrderItems are shipped.
 */
export function orderItemsAreShipped(order: Order) {
    return getOrderItems(order)
        .filter(orderItem => !orderItem.cancelled)
        .every(isShipped);
}

/**
 * Returns true if all OrderItems in the order are cancelled
 */
export function orderItemsAreAllCancelled(order: Order) {
    return getOrderItems(order).every(orderItem => orderItem.cancelled);
}

function getOrderItems(order: Order): OrderItem[] {
    return order.lines.reduce((orderItems, line) => [...orderItems, ...line.items], [] as OrderItem[]);
}
function getNonCancelledItems(order: Order): OrderItem[] {
    return getOrderItems(order).filter(orderItem => !orderItem.cancelled);
}

function isDelivered(orderItem: OrderItem) {
    return orderItem.fulfillment?.state === 'Delivered';
}
function isShipped(orderItem: OrderItem) {
    return orderItem.fulfillment?.state === 'Shipped';
}
