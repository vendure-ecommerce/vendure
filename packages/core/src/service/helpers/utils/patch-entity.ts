import { VendureEntity } from '../../../entity/base/base.entity';

export type InputPatch<T> = { [K in keyof T]?: T[K] | null };

/**
 * Updates only the specified properties from an Input object as long as the value is not
 * undefined. Null values can be passed, however, which will set the corresponding entity
 * field to "null". So care must be taken that this is only done on nullable fields.
 *
 * **Careful: This function returns the same passed in entity, which may have been mutated!**
 *
 * The mutation gets relevant if you bind the return value to a separate variable, then you have
 * two references that may mutate the same object. Example:
 *
 * ```js
 * const admin1 = new Administrator(...);
 * const admin2 = patchEntity(admin1, input);
 * // Changing fields on `admin2` will now also affect `admin1`!
 * // This can be dangerous, use this function responsibly.
 * ```
 */
export function patchEntity<T extends VendureEntity, I extends InputPatch<T>>(entity: T, input: I): T {
    for (const key of Object.keys(entity)) {
        const value = input[key as keyof T];
        if (key === 'customFields' && value) {
            patchEntity((entity as any)[key], value as any);
        } else if (value !== undefined && key !== 'id') {
            entity[key as keyof T] = value as any;
        }
    }
    return entity;
}
