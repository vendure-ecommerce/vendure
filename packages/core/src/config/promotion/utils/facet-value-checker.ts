import { ID } from '@vendure/common/lib/shared-types';
import { unique } from '@vendure/common/lib/unique';
import { RequestContext } from '../../../api';

import { TtlCache } from '../../../common/ttl-cache';
import { idsAreEqual } from '../../../common/utils';
import { TransactionalConnection } from '../../../connection/transactional-connection';
import { OrderLine } from '../../../entity/order-line/order-line.entity';
import { ProductVariant } from '../../../entity/product-variant/product-variant.entity';

/**
 * @description
 * The FacetValueChecker is a helper class used to determine whether a given OrderLine consists
 * of ProductVariants containing the given FacetValues.
 *
 * @example
 * ```TypeScript
 * import { FacetValueChecker, LanguageCode, PromotionCondition, TransactionalConnection } from '\@vendure/core';
 *
 * let facetValueChecker: FacetValueChecker;
 *
 * export const hasFacetValues = new PromotionCondition({
 *   code: 'at_least_n_with_facets',
 *   description: [
 *     { languageCode: LanguageCode.en, value: 'Buy at least { minimum } products with the given facets' },
 *   ],
 *   args: {
 *     minimum: { type: 'int' },
 *     facets: { type: 'ID', list: true, ui: { component: 'facet-value-form-input' } },
 *   },
 *   init(injector) {
 *     facetValueChecker = new FacetValueChecker(injector.get(TransactionalConnection));
 *   },
 *   // tslint:disable-next-line:no-shadowed-variable
 *   async check(ctx, order, args) {
 *     let matches = 0;
 *     for (const line of order.lines) {
 *       if (await facetValueChecker.hasFacetValues(line, args.facets)) {
 *           matches += line.quantity;
 *       }
 *     }
 *     return args.minimum <= matches;
 *   },
 * });
 * ```
 *
 * @docsCategory Promotions
 */
export class FacetValueChecker {
    private variantCache = new TtlCache<ID, ProductVariant>({ ttl: 5000 });

    constructor(private connection: TransactionalConnection) {}
    /**
     * @description
     * Checks a given {@link OrderLine} against the facetValueIds and returns
     * `true` if the associated {@link ProductVariant} & {@link Product} together
     * have *all* the specified {@link FacetValue}s.
     */
    async hasFacetValues(orderLine: OrderLine, facetValueIds: ID[], ctx?: RequestContext): Promise<boolean> {
        let variant = this.variantCache.get(orderLine.productVariant.id);
        if (!variant) {
            variant = await this.connection
                .getRepository(ctx, ProductVariant)
                .findOne(orderLine.productVariant.id, {
                    relations: ['product', 'product.facetValues', 'facetValues'],
                });
            if (!variant) {
                return false;
            }
            this.variantCache.set(variant.id, variant);
        }
        const allFacetValues = unique([...variant.facetValues, ...variant.product.facetValues], 'id');
        return facetValueIds.reduce(
            (result, id) => result && !!allFacetValues.find(fv => idsAreEqual(fv.id, id)),
            true as boolean,
        );
    }
}
