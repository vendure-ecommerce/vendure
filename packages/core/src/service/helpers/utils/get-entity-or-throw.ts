import { ID, Type } from '@vendure/common/lib/shared-types';
import { Connection, FindOneOptions } from 'typeorm';

import { EntityNotFoundError } from '../../../common/error/errors';
import { ChannelAware, SoftDeletable } from '../../../common/types/common-types';
import { VendureEntity } from '../../../entity/base/base.entity';

import { findOneInChannel } from './channel-aware-orm-utils';

/**
 * Attempts to find an entity of the given type and id, and throws an error if not found.
 * If the entity is a ChannelAware type, then the `channelId` must be supplied or else
 * the function will "fail" by resolving to the `never` type.
 */
export async function getEntityOrThrow<T extends VendureEntity>(
    connection: Connection,
    entityType: Type<T>,
    id: ID,
    findOptions?: FindOneOptions<T>,
): Promise<T extends ChannelAware ? never : T>;
export async function getEntityOrThrow<T extends VendureEntity | ChannelAware>(
    connection: Connection,
    entityType: Type<T>,
    id: ID,
    channelId: ID,
    findOptions?: FindOneOptions<T>,
    eager?: boolean,
): Promise<T>;
export async function getEntityOrThrow<T extends VendureEntity>(
    connection: Connection,
    entityType: Type<T>,
    id: ID,
    findOptionsOrChannelId?: FindOneOptions<T> | ID,
    maybeFindOptions?: FindOneOptions<T>,
    eager?: boolean,
): Promise<T> {
    let entity: T | undefined;
    if (isId(findOptionsOrChannelId)) {
        entity = await findOneInChannel(
            connection,
            entityType,
            id,
            findOptionsOrChannelId,
            maybeFindOptions,
            eager,
        );
    } else {
        entity = await connection.getRepository(entityType).findOne(id, findOptionsOrChannelId);
    }
    if (!entity || (entity.hasOwnProperty('deletedAt') && (entity as T & SoftDeletable).deletedAt !== null)) {
        throw new EntityNotFoundError(entityType.name as any, id);
    }
    return entity;
}

function isId(value: unknown): value is ID {
    return typeof value === 'string' || typeof value === 'number';
}
