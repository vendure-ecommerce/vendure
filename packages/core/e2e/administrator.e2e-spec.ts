import { SUPER_ADMIN_USER_IDENTIFIER } from '@vendure/common/lib/shared-constants';
import { createTestEnvironment } from '@vendure/testing';
import { fail } from 'assert';
import gql from 'graphql-tag';
import path from 'path';
import { afterAll, beforeAll, describe, expect, it } from 'vitest';

import { initialData } from '../../../e2e-common/e2e-initial-data';
import { testConfig, TEST_SETUP_TIMEOUT_MS } from '../../../e2e-common/test-config';

import { ADMINISTRATOR_FRAGMENT } from './graphql/fragments';
import * as Codegen from './graphql/generated-e2e-admin-types';
import { AdministratorFragment, DeletionResult } from './graphql/generated-e2e-admin-types';
import { CREATE_ADMINISTRATOR, UPDATE_ADMINISTRATOR } from './graphql/shared-definitions';
import { assertThrowsWithMessage } from './utils/assert-throws-with-message';

describe('Administrator resolver', () => {
    const { server, adminClient } = createTestEnvironment(testConfig());
    let createdAdmin: AdministratorFragment;

    beforeAll(async () => {
        await server.init({
            initialData,
            productsCsvPath: path.join(__dirname, 'fixtures/e2e-products-minimal.csv'),
            customerCount: 1,
        });
        await adminClient.asSuperAdmin();
    }, TEST_SETUP_TIMEOUT_MS);

    afterAll(async () => {
        await server.destroy();
    });

    it('administrators', async () => {
        const result = await adminClient.query<
            Codegen.GetAdministratorsQuery,
            Codegen.GetAdministratorsQueryVariables
        >(GET_ADMINISTRATORS);
        expect(result.administrators.items.length).toBe(1);
        expect(result.administrators.totalItems).toBe(1);
    });

    it('createAdministrator', async () => {
        const result = await adminClient.query<
            Codegen.CreateAdministratorMutation,
            Codegen.CreateAdministratorMutationVariables
        >(CREATE_ADMINISTRATOR, {
            input: {
                emailAddress: 'test@test.com',
                firstName: 'First',
                lastName: 'Last',
                password: 'password',
                roleIds: ['1'],
            },
        });

        createdAdmin = result.createAdministrator;
        expect(createdAdmin).toMatchSnapshot();
    });

    it('administrator', async () => {
        const result = await adminClient.query<
            Codegen.GetAdministratorQuery,
            Codegen.GetAdministratorQueryVariables
        >(GET_ADMINISTRATOR, {
            id: createdAdmin.id,
        });
        expect(result.administrator).toEqual(createdAdmin);
    });

    it('updateAdministrator', async () => {
        const result = await adminClient.query<
            Codegen.UpdateAdministratorMutation,
            Codegen.UpdateAdministratorMutationVariables
        >(UPDATE_ADMINISTRATOR, {
            input: {
                id: createdAdmin.id,
                emailAddress: 'new-email',
                firstName: 'new first',
                lastName: 'new last',
                password: 'new password',
                roleIds: ['2'],
            },
        });
        expect(result.updateAdministrator).toMatchSnapshot();
    });

    it('updateAdministrator works with partial input', async () => {
        const result = await adminClient.query<
            Codegen.UpdateAdministratorMutation,
            Codegen.UpdateAdministratorMutationVariables
        >(UPDATE_ADMINISTRATOR, {
            input: {
                id: createdAdmin.id,
                emailAddress: 'newest-email',
            },
        });
        expect(result.updateAdministrator.emailAddress).toBe('newest-email');
        expect(result.updateAdministrator.firstName).toBe('new first');
        expect(result.updateAdministrator.lastName).toBe('new last');
    });

    it(
        'updateAdministrator throws with invalid roleId',
        assertThrowsWithMessage(
            () =>
                adminClient.query<
                    Codegen.UpdateAdministratorMutation,
                    Codegen.UpdateAdministratorMutationVariables
                >(UPDATE_ADMINISTRATOR, {
                    input: {
                        id: createdAdmin.id,
                        emailAddress: 'new-email',
                        firstName: 'new first',
                        lastName: 'new last',
                        password: 'new password',
                        roleIds: ['999'],
                    },
                }),
            'No Role with the id "999" could be found',
        ),
    );

    it('deleteAdministrator', async () => {
        const { administrators: before } = await adminClient.query<
            Codegen.GetAdministratorsQuery,
            Codegen.GetAdministratorsQueryVariables
        >(GET_ADMINISTRATORS);
        expect(before.totalItems).toBe(2);

        const { deleteAdministrator } = await adminClient.query<
            Codegen.DeleteAdministratorMutation,
            Codegen.DeleteAdministratorMutationVariables
        >(DELETE_ADMINISTRATOR, {
            id: createdAdmin.id,
        });

        expect(deleteAdministrator.result).toBe(DeletionResult.DELETED);

        const { administrators: after } = await adminClient.query<
            Codegen.GetAdministratorsQuery,
            Codegen.GetAdministratorsQueryVariables
        >(GET_ADMINISTRATORS);
        expect(after.totalItems).toBe(1);
    });

    it('cannot delete sole SuperAdmin', async () => {
        const { administrators: before } = await adminClient.query<
            Codegen.GetAdministratorsQuery,
            Codegen.GetAdministratorsQueryVariables
        >(GET_ADMINISTRATORS);
        expect(before.totalItems).toBe(1);
        expect(before.items[0].emailAddress).toBe('superadmin');

        try {
            const { deleteAdministrator } = await adminClient.query<
                Codegen.DeleteAdministratorMutation,
                Codegen.DeleteAdministratorMutationVariables
            >(DELETE_ADMINISTRATOR, {
                id: before.items[0].id,
            });
            fail('Should have thrown');
        } catch (e: any) {
            expect(e.message).toBe('The sole SuperAdmin cannot be deleted');
        }

        const { administrators: after } = await adminClient.query<
            Codegen.GetAdministratorsQuery,
            Codegen.GetAdministratorsQueryVariables
        >(GET_ADMINISTRATORS);
        expect(after.totalItems).toBe(1);
    });

    it(
        'cannot remove SuperAdmin role from sole SuperAdmin',
        assertThrowsWithMessage(async () => {
            const result = await adminClient.query<
                Codegen.UpdateAdministratorMutation,
                Codegen.UpdateAdministratorMutationVariables
            >(UPDATE_ADMINISTRATOR, {
                input: {
                    id: 'T_1',
                    roleIds: [],
                },
            });
        }, 'Cannot remove the SuperAdmin role from the sole SuperAdmin'),
    );

    it('cannot query a deleted Administrator', async () => {
        const { administrator } = await adminClient.query<
            Codegen.GetAdministratorQuery,
            Codegen.GetAdministratorQueryVariables
        >(GET_ADMINISTRATOR, {
            id: createdAdmin.id,
        });

        expect(administrator).toBeNull();
    });

    it('activeAdministrator', async () => {
        await adminClient.asAnonymousUser();

        const { activeAdministrator: result1 } = await adminClient.query<Codegen.ActiveAdministratorQuery>(
            GET_ACTIVE_ADMINISTRATOR,
        );
        expect(result1).toBeNull();

        await adminClient.asSuperAdmin();

        const { activeAdministrator: result2 } = await adminClient.query<Codegen.ActiveAdministratorQuery>(
            GET_ACTIVE_ADMINISTRATOR,
        );
        expect(result2?.emailAddress).toBe(SUPER_ADMIN_USER_IDENTIFIER);
    });

    it('updateActiveAdministrator', async () => {
        const { updateActiveAdministrator } = await adminClient.query<
            Codegen.UpdateActiveAdministratorMutation,
            Codegen.UpdateActiveAdministratorMutationVariables
        >(UPDATE_ACTIVE_ADMINISTRATOR, {
            input: {
                firstName: 'Thomas',
                lastName: 'Anderson',
                emailAddress: 'neo@metacortex.com',
            },
        });

        expect(updateActiveAdministrator.firstName).toBe('Thomas');
        expect(updateActiveAdministrator.lastName).toBe('Anderson');

        const { activeAdministrator } = await adminClient.query<Codegen.ActiveAdministratorQuery>(
            GET_ACTIVE_ADMINISTRATOR,
        );

        expect(activeAdministrator?.firstName).toBe('Thomas');
        expect(activeAdministrator?.user.identifier).toBe('neo@metacortex.com');
    });
});

export const GET_ADMINISTRATORS = gql`
    query GetAdministrators($options: AdministratorListOptions) {
        administrators(options: $options) {
            items {
                ...Administrator
            }
            totalItems
        }
    }
    ${ADMINISTRATOR_FRAGMENT}
`;

export const GET_ADMINISTRATOR = gql`
    query GetAdministrator($id: ID!) {
        administrator(id: $id) {
            ...Administrator
        }
    }
    ${ADMINISTRATOR_FRAGMENT}
`;

export const GET_ACTIVE_ADMINISTRATOR = gql`
    query ActiveAdministrator {
        activeAdministrator {
            ...Administrator
        }
    }
    ${ADMINISTRATOR_FRAGMENT}
`;

export const UPDATE_ACTIVE_ADMINISTRATOR = gql`
    mutation UpdateActiveAdministrator($input: UpdateActiveAdministratorInput!) {
        updateActiveAdministrator(input: $input) {
            ...Administrator
        }
    }
    ${ADMINISTRATOR_FRAGMENT}
`;

export const DELETE_ADMINISTRATOR = gql`
    mutation DeleteAdministrator($id: ID!) {
        deleteAdministrator(id: $id) {
            message
            result
        }
    }
`;
