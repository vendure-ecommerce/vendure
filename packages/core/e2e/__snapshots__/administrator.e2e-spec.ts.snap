// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Administrator resolver createAdministrator 1`] = `
Object {
  "emailAddress": "test@test.com",
  "firstName": "First",
  "id": "T_2",
  "lastName": "Last",
  "user": Object {
    "id": "T_3",
    "identifier": "test@test.com",
    "lastLogin": null,
    "roles": Array [
      Object {
        "code": "__super_admin_role__",
        "description": "SuperAdmin",
        "id": "T_1",
        "permissions": Array [
          "Authenticated",
          "SuperAdmin",
          "UpdateGlobalSettings",
          "CreateCatalog",
          "ReadCatalog",
          "UpdateCatalog",
          "DeleteCatalog",
          "CreateSettings",
          "ReadSettings",
          "UpdateSettings",
          "DeleteSettings",
          "CreateAdministrator",
          "ReadAdministrator",
          "UpdateAdministrator",
          "DeleteAdministrator",
          "CreateAsset",
          "ReadAsset",
          "UpdateAsset",
          "DeleteAsset",
          "CreateChannel",
          "ReadChannel",
          "UpdateChannel",
          "DeleteChannel",
          "CreateCollection",
          "ReadCollection",
          "UpdateCollection",
          "DeleteCollection",
          "CreateCountry",
          "ReadCountry",
          "UpdateCountry",
          "DeleteCountry",
          "CreateCustomer",
          "ReadCustomer",
          "UpdateCustomer",
          "DeleteCustomer",
          "CreateCustomerGroup",
          "ReadCustomerGroup",
          "UpdateCustomerGroup",
          "DeleteCustomerGroup",
          "CreateFacet",
          "ReadFacet",
          "UpdateFacet",
          "DeleteFacet",
          "CreateOrder",
          "ReadOrder",
          "UpdateOrder",
          "DeleteOrder",
          "CreatePaymentMethod",
          "ReadPaymentMethod",
          "UpdatePaymentMethod",
          "DeletePaymentMethod",
          "CreateProduct",
          "ReadProduct",
          "UpdateProduct",
          "DeleteProduct",
          "CreatePromotion",
          "ReadPromotion",
          "UpdatePromotion",
          "DeletePromotion",
          "CreateShippingMethod",
          "ReadShippingMethod",
          "UpdateShippingMethod",
          "DeleteShippingMethod",
          "CreateTag",
          "ReadTag",
          "UpdateTag",
          "DeleteTag",
          "CreateTaxCategory",
          "ReadTaxCategory",
          "UpdateTaxCategory",
          "DeleteTaxCategory",
          "CreateTaxRate",
          "ReadTaxRate",
          "UpdateTaxRate",
          "DeleteTaxRate",
          "CreateSystem",
          "ReadSystem",
          "UpdateSystem",
          "DeleteSystem",
          "CreateZone",
          "ReadZone",
          "UpdateZone",
          "DeleteZone",
        ],
      },
    ],
  },
}
`;

exports[`Administrator resolver updateAdministrator 1`] = `
Object {
  "emailAddress": "new-email",
  "firstName": "new first",
  "id": "T_2",
  "lastName": "new last",
  "user": Object {
    "id": "T_3",
    "identifier": "new-email",
    "lastLogin": null,
    "roles": Array [
      Object {
        "code": "__customer_role__",
        "description": "Customer",
        "id": "T_2",
        "permissions": Array [
          "Authenticated",
        ],
      },
    ],
  },
}
`;
